doc/design/model_zoo.md

```diff
@@ -0,0 +1,174 @@
+
+1. A collection of model definitions is a Git repository of source files. 
+1. To describe dependencies, we require a Dockerfile at the root directory of the repository.
+1. To release a repository, we checkout the specific version and run `docker run` with the Dockerfile.
+1. Each release has a globally unique ID (GUID) represented by the hash ID of the Docker image.
+1. The Docker image must support a protocol of command-line arguments so that SQLFlow can pass in required information like hyperparameters from the `docker run` command.
+
+### Submitter Programs
+
+For a SELECT program using the SQLFlow syntax extension, the SQLFlow server converts it into a *submitter* program, usually in Python.
+
+Consider the training statement at the beginning of this document. Suppose that the Python class `MyDNNRegressor` is in source file `my_dnn_regressor.py` in the repository https://github.com/a_data_scientist/regressors, which builds into Docker image `a_data_scientist/regressors`.  Let us also assume that the submitter program is `/var/sqlflow/submitters/an_analyst/my_first_model.py`.  SQLFlow can run the submitter using the Docker image as follows:
+
+```bash
+docker run --rm -it \
+  -v /var/sqlflow/submitters:/submitters \
+  a_data_scientist/regressors \
+    python /submitters/an_analyst/my_first_model.py
+```
+
+If we run SQLFlow in a Docker container, to allow it to run another Docker container, we must enable the Docker-in-Docker feature, say, following [this blog post](https://itnext.io/docker-in-docker-521958d34efd).  Suppose that the SQLFlow server runs in a Kubernetes cluster, where we cannot enable Docker-in-Docker. In such a case, we can make the SQLFlow server call Kubernetes API to run a Pod that executes the above `docker run` command line.
+
+The training submitter program `my_first_model.py`, running in the model definition Docker container, should be able to submit a (distributed) training job to a preconfigured cluster.  Once the job completes, the submitter program adds, or edits, a row in a **model zoo table**.
+
+### Model Zoo Data Schema
+
+The model zoo table is in a database deployed as part of the SQLFlow service. This database might not be the one that holds data sources.  The only requirement of the model zoo table is to have a particular data schema that contains at least the following fields.
+
+1. model ID (key), specified by the INTO clause, or `an_analyst/my_first_model` in the above example.
+1. Docker image ID, the Docker commit ID of the image `a_data_scientist/regressors` in the above example.
+1. submitter program, the source code of the submitter program, `my_first_model.py` in the above example, or its MD5 hash.
+1. data converter, the COLUMN and LABEL clauses.
```

Comment:
It seems we should save the extended SQL itself instead of the two above fields, for example: if the user trained two different datasets at different time (daily analysis e. g.), the model zoo should know which model comes from which dataset. 
As a supplement, the SQLFlow version and training date/time can be present too, just like the build number in software versioning.
At last, there should be a description field that records training result (test accuracy etc.) by default, or users cannot distingish different models if they have too many.

Reply:
I agree that we might need more fields here. Each exists for some reason. We could add these fields together with the reasons later.

Reply:
We should probably save both the training SQL statement as well as the "data converter", because we are doing "feature derivation" when training, the "data converter" is generated based on the training data. If the data used for predict is of different format of training data, go through "feature derivation" again might cause some error.